admin_terraform_version.go

```diff
@@ -0,0 +1,203 @@
+// Note that admin terraform versions are only available in Terraform Enterprise.
+//
+// TFE API docs: https://www.terraform.io/docs/cloud/api/admin/terraform-versions.html
+type AdminTerraformVersions interface {
+	// List all the terraform versions.
+	List(ctx context.Context, options AdminTerraformVersionsListOptions) (*AdminTerraformVersionsList, error)
+
+	// Read a terraform version by its ID.
+	Read(ctx context.Context, id string) (*AdminTerraformVersion, error)
+
+	// Create a terraform version.
+	Create(ctx context.Context, options AdminTerraformVersionCreateOptions) (*AdminTerraformVersion, error)
+
+	// Update a terraform version.
+	Update(ctx context.Context, id string, options AdminTerraformVersionUpdateOptions) (*AdminTerraformVersion, error)
+
+	// Delete a terraform version
+	Delete(ctx context.Context, id string) error
+}
+
+// adminTerraformVersions implements AdminTerraformVersions.
+type adminTerraformVersions struct {
+	client *Client
+}
+
+// AdminTerraformVersion represents a Terraform Version
+type AdminTerraformVersion struct {
+	ID        string `jsonapi:"primary,terraform-versions"`
+	Version   string `jsonapi:"attr,version"`
+	URL       string `jsonapi:"attr,url"`
+	Sha       string `jsonapi:"attr,sha"`
+	Official  bool   `jsonapi:"attr,official"`
+	Enabled   bool   `jsonapi:"attr,enabled"`
+	Beta      bool   `jsonapi:"attr,beta"`
+	Usage     int    `jsonapi:"attr,usage"`
+	CreatedAt string `jsonapi:"attr,created-at"`
+}
+
+// AdminTerraformVersionsListOptions represents the options for listing
+// terraform versions.
+type AdminTerraformVersionsListOptions struct {
+	ListOptions
+}
+
+// AdminTerraformVersionsList represents a list of terraform versions.
+type AdminTerraformVersionsList struct {
+	*Pagination
+	Items []*AdminTerraformVersion
+}
+
+// List all the terraform versions.
+func (a *adminTerraformVersions) List(ctx context.Context, options AdminTerraformVersionsListOptions) (*AdminTerraformVersionsList, error) {
+	req, err := a.client.newRequest("GET", "admin/terraform-versions", &options)
+	if err != nil {
+		return nil, err
+	}
+
+	awl := &AdminTerraformVersionsList{}
+	err = a.client.do(ctx, req, awl)
+	if err != nil {
+		return nil, err
+	}
+
+	return awl, nil
+}
+
+// Read a terraform version by its ID.
+func (a *adminTerraformVersions) Read(ctx context.Context, id string) (*AdminTerraformVersion, error) {
+	if !validStringID(&id) {
+		return nil, ErrInvalidTerraformVersionID
+	}
+
+	u := fmt.Sprintf("admin/terraform-versions/%s", url.QueryEscape(id))
+	req, err := a.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	tfv := &AdminTerraformVersion{}
+	err = a.client.do(ctx, req, tfv)
+	if err != nil {
+		return nil, err
+	}
+
+	return tfv, nil
+}
+
+// AdminTerraformVersionCreateOptions for creating a terraform version.
+// https://www.terraform.io/docs/cloud/api/admin/terraform-versions.html#request-body
+type AdminTerraformVersionCreateOptions struct {
```

Comment:
Every CreateOptions type in this project has a `ID` field (`string`) "for internal use only!"

...I've never actually have a clue what it's for, though, so I went on a spelunking session and [found when this was added and discussion around it](https://github.com/hashicorp/go-tfe/pull/1#discussion_r200041679)

Can you dig in to that and report back if that actually matters anymore? I don't immediately see why `jsonapi` needs this field at all. If it's not clear and we want to just move forward, I'd just add it for consistency's sake and we can revisit separately and remove them _all_ if it makes sense (it's always been marked as internal private API only that we can totally rip out if it doesn't make sense).

Reply:
Here are a few findings.

1. Checked a number of resources (Organizations, Variables, Workspaces, StateVersions) and when we create those resources, we do *not* utilize the `id` of the request body.
2. The *only* requirement for creating a [resource according to the jsonapi spec](https://jsonapi.org/format/#crud-creating) is a *type* field.
3. The only time `id`s are being used is when the [client wants to generate a unique identifier](https://jsonapi.org/format/#crud-creating-client-ids) for that resource.

**Recommendation**

Since we are *not* using the `id` field internally as part of creating resources, and they are *not* required by JSON:API, I'd recommend we remove them all. I can create a ticket for this. And seeing how this followup ticket would be a quick effort, are you strongly in favor of adding it here for the sake of consistency even though it will be removed shortly afterwords? I'm not sure it adds any consistency-value.

Reply:
Ahh nice work digging in to this. 

> they are not required by JSON:API

They are, aren't they? Don't you still need some sort of public field to add the fieldtag that makes the type be sent properly? Or are you saying you see a way where we can do that, just _not_ as a part of `CreateOptions`/`UpdateOptions`? 

> are you strongly in favor of adding it here

Nope, whatever makes sense to you! Doesn't matter.

Reply:
> They are, aren't they? Don't you still need some sort of public field to add the fieldtag

Yes, that is correct. The requirement is that there is some struct attribute that is the place where the `json:primary,<resource>` resides. 

I'm going to make it `Type` and go with that. That will also answer the issue below about validating type. 

Also going to make a ticket in Asana to come back and fix this id/type stuff for all the other APIs.